lib/compiler/src/compile.erl

```diff
@@ -1386,17 +1437,33 @@ encrypt({des3_cbc=Type,Key,IVec,BlockSize}, Bin0) ->
 save_core_code(Code, St) ->
     {ok,Code,St#compile{core_code=cerl:from_records(Code)}}.
 
-beam_asm(Code0, #compile{ifile=File,abstract_code=Abst,
-			 options=CompilerOpts,mod_options=Opts0}=St) ->
+beam_asm(Code0, #compile{ifile=File, types=Types, specs=Specs0,
+                         records=Records, abstract_code=Abst,
+			 options=CompilerOpts, mod_options=Opts0}=St) ->
     Source = paranoid_absname(File),
     Opts1 = lists:map(fun({debug_info_key,_}) -> {debug_info_key,'********'};
 			 (Other) -> Other
 		      end, Opts0),
     Opts2 = [O || O <- Opts1, effects_code_generation(O)],
-    case beam_asm:module(Code0, Abst, Source, Opts2, CompilerOpts) of
+    {Mod,Exp0,_,_,_} = Code0,
+    Specs = filter_specs(Specs0, Mod, Exp0),
+    case beam_asm:module(Code0, Types, Specs, Records, Abst, Source, Opts2, CompilerOpts) of
```

Comment:
What if instead of changing `beam_asm` to receive `Types, Specs, Records, Abst`, you pass a list of extra chunks:

```erlang
ExtraChunks = [{<<"Type">>, Types}, {<<"Spec">>, Spec}, ...}]
beam_asm:module(Code0, ExtraChunks, Source, Opts2, CompilerOpts)
```

Then we will no longer need to change `beam_asm` every time we want to add a new optional chunk related to the erlang abstract format.